### PR TITLE
[MPS] Add support for autocast in MPS 

### LINKED ISSUE #88415: Enable AMP for MPS devices
### 🚀 The feature, motivation and pitch

https://github.com/pytorch/pytorch/issues/78168 States that fp16 support for mps devices in general should be possible, but autocas only works with `cpu` and `cuda` device types. When enabling it manually, on mps it does not show any additional conversions:

```python
>>> torch.set_autocast_enabled(True)
>>> with capture_logs(is_mode=True) as logs, LoggingTensorMode():
...     a = torch.rand(10, 10, dtype=torch.float, device='mps')
...     b = torch.rand(10, 10, dtype=torch.float, device='mps')
...     c = torch.addmm(a, a, b)
>>> for l in logs:
...     print(l)
```

prints the following:
```
$0 = torch._ops.aten.rand.default([10, 10], dtype=torch.float32, device=device(type='mps'), pin_memory=False)
$1 = torch._ops.aten.rand.default([10, 10], dtype=torch.float32, device=device(type='mps'), pin_memory=False)
$2 = torch._ops.aten.addmm.default($0, $0, $1)
```

given the fp16 support, it would be nice to have autocast and amp in general working on MPS devices as well

### Alternatives

_No response_

### Additional context

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5 @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev as per our discussion on slack.

### PR DESCRIPTION
**Pull request by kulinseth**
Fixes https://github.com/pytorch/pytorch/issues/88415

Need to run inductor/test_cpu_select_algorithm

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -62,8 +62,8 @@ def set_device_states(devices, states) -> None:
             torch.cuda.set_rng_state(state)
 
 def _get_autocast_kwargs():
-    gpu_autocast_kwargs = {"enabled": torch.is_autocast_enabled(),
-                           "dtype": torch.get_autocast_gpu_dtype(),
+    gpu_autocast_kwargs = {"enabled": torch.is_autocast_enabled() or torch.is_autocast_mps_enabled(),
+                           "dtype": torch.get_autocast_mps_dtype() if torch.backends.mps.is_available() else torch.get_autocast_gpu_dtype(),
```
Comment (albanD) [inline torch/utils/checkpoint.py:66]: Don't we have a device-independent method to access the current dtype?

**Review by albanD — COMMENTED** [review]
Sounds pretty good. Mostly questions about testing.

**Comment by johnrichardrinehart** [comment]
@kulinseth It seems there is still some work to be done, here. What's your plan to address some of these comments?

**Comment by albanD** [comment]
Any update on this @kulinseth

**Comment by kulinseth** [comment]
> Any update on this @kulinseth ?

Yes, I need to rebase and update here. I will do it this week.

**Comment by johnrichardrinehart** [comment]
@kulinseth Still in progress? Looks so based on #104191 .

**Comment by djollieb** [comment]
@kulinseth Is there any update on this please? This change would fix so many issues. Thank you

**Comment by kulinseth** [comment]
> @kulinseth Is there any update on this please? This change would fix so many issues. Thank you

Yes , I will cleanup and rebase again . Will have something by early next week to try

**Comment by johnrichardrinehart** [comment]
> > @kulinseth Is there any update on this please? This change would fix so many issues. Thank you
> 
> Yes , I will cleanup and rebase again . Will have something by early next week to try

Any updates @kulinseth ?

**Comment by diffractometer** [comment]
@kulinseth I'll buy you a coffee!

**Comment by john1089** [comment]
Any updates @kulinseth ?

**Comment by PhilippeFerreiraDeSousa** [comment]
@kulinseth Can we do something to help?

**Comment by larsrnielsen** [comment]
@kulinseth any chance it will be this year?

**Comment by fakerybakery** [comment]
Any updates @kulinseth it'd be really great to see this PR merged and would greatly improve so many upstream projects!

**Comment by Abdullah-R** [comment]
Any updates @kulinseth? Have been eagerly awaiting this merge, would be super useful for various projects!

**Comment by larsrnielsen** [comment]
@kulinseth as you have stopped to work on it (according to your activity), is there someone else that can get it done?

**Comment by kulinseth** [comment]
> @kulinseth as you have stopped to work on it (according to your activity), is there someone else that can get it done?

I am enlisting help from others (cc @malfet ) on this if we I can’t get to it in next 2 weeks .

**Comment by rajdtta** [comment]
@kulinseth just wanted to follow up on the progress of this. 

Appreciate the work you're doing :)

**Comment by john1089** [comment]
So when this pr will be merged?

**Comment by larsrnielsen** [comment]
> > @kulinseth as you have stopped to work on it (according to your activity), is there someone else that can get it done?
> 
> I am enlisting help from others (cc @malfet ) on this if we I can’t get to it in next 2 weeks .

any updates?

**Comment by piotrpersona** [comment]
any updates on this?

**Comment by helsenIgnace** [comment]
When do you thing this pr could be merged?

**Comment by Abdullah-R** [comment]
Is progress being made or has this PR been abandoned?

**Comment by petros94** [comment]
An update pls? 🙂

**Comment by galenmarchetti** [comment]
also pinging for updates, would love to see this

**Comment by bastienjalbert** [comment]
Will this PR merged one day ?

**Comment by bghira** [comment]
@albanD please re-review

**Comment by bghira** [comment]
@kulinseth a gentle ping since pytorch 2.3 has bf16 on mps now

**Comment by arodri48** [comment]
@kulinseth I am interested in trying out AMP on my M3 Pro. Are there any updates regarding this PR?

**Comment by janboeye** [comment]
hi, @kulinseth 
I rebase your patch on latest main branch, could you help to review [it](https://github.com/janboeye/pytorch/commit/49becef211db7728af95bd3c965ff60162fb93a6) ?

Thanks

**Review by malfet — APPROVED** [review]
LGTM

**Comment by benjaminhuo** [comment]
closed instead of merged?

**Comment by qqaatw** [comment]
> closed instead of merged?

It is merged. Close is simply a side effect of merging in PyTorch repository.

**Comment by jeanschmidt** [comment]
I believe this PR introduced the regressions:

https://github.com/pytorch/pytorch/actions/runs/9753683022/job/26920813743#step:20:1325
https://github.com/pytorch/pytorch/actions/runs/9753683022/job/26920813614#step:20:1332
https://github.com/pytorch/pytorch/actions/runs/9753683022/job/26920813500#step:20:2948

So reverting in order to clarify if hey will be fixed.

**Comment by skotapati** [comment]
Hm, didn't see any test failures earlier. Rebased & rerunning the tests

File: torch/cuda/amp/common.py [hunk 2]
```diff
@@ -7,4 +7,8 @@
 
 
 def amp_definitely_not_available():
-    return not (torch.cuda.is_available() or find_spec("torch_xla"))
+    return not (
```
Comment (albanD) [inline torch/cuda/amp/common.py:10]: Why do you change this function? I would expect only the one from torch.amp is used and should work fine.

Comment (skotapati) [reply torch/cuda/amp/common.py:10]: I see, I didn't realize this function wasn't called.

**Comment by skotapati** [comment]
@malfet @jeanschmidt There appears to be an infrastructure issue of some kind, I'm getting `/home/ec2-user/actions-runner/externals/node20/bin/node: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by /home/ec2-user/actions-runner/externals/node20/bin/node)`

I've tried relaunching the tests a few times, but no luck. Any recommendations for how to fix this?

**Comment by jjuliano** [comment]
> @malfet @jeanschmidt There appears to be an infrastructure issue of some kind, I'm getting `/home/ec2-user/actions-runner/externals/node20/bin/node: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by /home/ec2-user/actions-runner/externals/node20/bin/node)`
> 
> I've tried relaunching the tests a few times, but no luck. Any recommendations for how to fix this?

I'm also rooting for this @kulinseth & @skotapati. Thank you for your work!

As per the problem:

It seems that GHA checkout@v3 uses a different nodejs version. I read the online solutions  here (https://github.com/actions/checkout/issues/1809) which says to add the following env vars:

```
export ACTIONS_RUNNER_FORCED_INTERNAL_NODE_VERSION=node16
export ACTIONS_RUNNER_FORCE_ACTIONS_NODE_VERSION=node16
```

and this declaration also solve the issue in most people.

```
ACTIONS_ALLOW_USE_UNSECURE_NODE_VERSION=true
```

**Comment by skotapati** [comment]
@clee2000 Thanks for looking into this. It looks like the failing tests have been reclassified as flaky on trunk, so I'm going to go ahead and reattempt the merge, if that's ok.

**Comment by clee2000** [comment]
Dr. CI is wrong, those should not be marked as flaky.  If you click on the link for the similar failure, it links to a commit that was red due to this PR

**Comment by albanD** [comment]
@clee2000 any insight on how to get more information on this given the error above is a rather cryptic "OSError: [Errno 9] Bad file descriptor" ?

**Comment by clee2000** [comment]
Given these are linux job, you could try sshing into the machine to debug in the CI environment if its not replicable locally.  Other than that I don't think I'll be of much help here

**Comment by malfet** [comment]
Looks like the failure is the same as it was when PR was reverted for a month ago? Just curious, what have changed since than?

**Comment by skotapati** [comment]
> Looks like the failure is the same as it was when PR was reverted for a month ago? Just curious, what have changed since than?

Just https://github.com/pytorch/pytorch/pull/99272/commits/8a166aef8042717e00af18f03fbee865fb397864 & a few rebases

> Given these are linux job, you could try sshing into the machine to debug in the CI environment if its not replicable locally. Is there anything the tests do involving printing, stdout/err, process exiting that might mess with testing? Testing in CI doesn't actually print to stdout or stderr, it prints to a file, which is then printed afterwards if the test fails. Maybe something is expecting a file descriptor for stdout/err but getting a file descriptor for an actual file is causing an error?

ssh to the machine seems like the best bet, how do I get access? @clee2000 @albanD

**Comment by malfet** [comment]
Anyone knows what has recently changed with the PR, as it seems to be passing the old failures, but now it fails some of MPS tests, which seems new

File: test/test_mps.py [hunk 3]
```diff
@@ -1203,6 +1203,29 @@ def __exit__(self, exec_type, exec_value, traceback):
...
+        with torch.autocast(device_type="mps"):
+            autocast_output_tensor = torch.mm(autocast_tensor_A, autocast_tensor_B)
+            autocast_output_tensor = torch.mm(autocast_tensor_A, autocast_output_tensor)
+
+        output_tensor = torch.mm(tensor_A, tensor_B)
+        output_tensor = torch.mm(tensor_A, output_tensor)
+
+        self.assertEqual(autocast_output_tensor.dtype, torch.float16, "Autocast output tensor was not expected type float16")
+        self.assertEqual(autocast_output_tensor,
+                         output_tensor.to(torch.float16),
```
Comment (hvaara) [inline test/test_mps.py:1225]: Is this expected to be `bfloat16` after https://github.com/pytorch/pytorch/pull/99272/commits/f282bf47427ba9b876b2ae496716921482d0e870?

Comment (malfet) [inline test/test_mps.py:1225]:
```suggestion
                         output_tensor.to(torch.bfloat16),
```

File: test/test_mps.py [hunk 4]
```diff
@@ -1203,6 +1203,29 @@ def __exit__(self, exec_type, exec_value, traceback):
...
+                         f"Autocast & non-autocast tensors did not match, \
+                         got:\n{autocast_output_tensor} \n{output_tensor.to(torch.float16)}")
```
Comment (malfet) [inline test/test_mps.py:1227]:
```suggestion
                         got:\n{autocast_output_tensor} \n{output_tensor.to(torch.bfloat16)}")
```

File: test/test_mps.py (hunk 3 above)
Comment (malfet) [reply test/test_mps.py]: Hmm, but than autocast is only available on MacOS14 or newer, is that expected?

Comment (hvaara) [reply test/test_mps.py]: Yes. I don't know if this is intentional or not.

cc @skotapati @kulinseth

Comment (skotapati) [reply test/test_mps.py]: Yeah sorry, missed updating the tests after switching to bf16

File: test/test_autocast.py [hunk 5]
```diff
@@ -344,6 +344,55 @@ def test_cache_disabled(self):
...
+        data = torch.randn(2, 3).to("mps")
+        weight = torch.nn.Parameter(torch.randn(4, 3).to("mps"))
+        weight_dtype_cast_counter = 0
+
+        class WeightDTypeCastCounterMode(TorchDispatchMode):
+            def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+                if (
+                    func is torch.ops.aten._to_copy.default
+                    and args[0] is weight
+                    and kwargs["dtype"] is torch.float16
```
Comment (hvaara) [inline test/test_autocast.py:373]:
```suggestion
                    and kwargs["dtype"] is torch.bfloat16
```

**Review by malfet — CHANGES_REQUESTED** [review]
This CI signal looks much better, but I'm a bit confused about why autocast should be done for bf16 type now, instead of fp16 one? Later makes much more sense, isn't it?

File: test/test_autocast.py (hunk 5 above)
Comment (skotapati) [reply test/test_autocast.py:373]: Ah shoot, thanks for catching this

**Comment by skotapati** [comment]
> This CI signal looks much better, but I'm a bit confused about why autocast should be done for bf16 type now, instead of fp16 one? Later makes much more sense, isn't it?

My original intention was to merge this PR and then add bf16 support in a separate one. However that would mean retroactively removing autocast support from MacOS 13 systems, so I thought it would be better to just stick with bf16 for MacOS14+ instead of creating inconsistency

@malfet Let me know what you think, I can switch back to fp16, and follow up with a separate bf16 PR if that would be better

**Comment by skotapati** [comment]
After further discussion with the rest of the team, switching back to fp16 for now in order to maintain performance on older systems

**Comment by bghira** [comment]
can bf16 remain as an option, please? we're expecting that on newer platforms with more sensitive training that requires bf16's range

**Review by malfet — APPROVED** [review]
Looks good to me

**Comment by malfet** [comment]
> can bf16 remain as an option, please? 

@bghira can you please elaborate why/when you need something like that? But if this is the case, we definitely should extend it in follow-up PR, but perhaps rules should be different that for fp16, aren't they?

File: test/test_mps.py (hunk 3 above)
Comment (defertoexpertise) [reply test/test_mps.py:1225]: Question trying to understand whats exactly going on with mps and bf16, saw on sonoma newer m2/m3 support bf16, but with flux and some other models for training was told bf16 isnt supported on macbooks, is it just that we need newer editions of pytorch and the devs from those trainers were mistaken that bf16 is supported?

Comment (bghira) [reply test/test_mps.py:1225]: m2+ hardware is needed and pytorch before 2.4 didn't have bf16 and pytorch after 2.4 has correctness issues on MPS. it's a sore situation and i think the safest answer for now is "don't use it"

**Comment by simepy** [comment]
@malfet @kulinseth Is this PR merged ?, looks like not, any advice on how to fix it ?
It will be sad to forget it after one year of work...

**Comment by bastienjalbert** [comment]
@malfet, @skotapati why the PR has been closed ?

**Comment by hvaara** [comment]
> @malfet, @skotapati why the PR has been closed ?

@bastienjalbert The PR is closed because merging code in PyTorch is more complicated than GitHub's default workflow and is handled by bot commands. If you're interested in learning more about what goes on in the process you can take a look at the [workflow file](https://github.com/pytorch/pytorch/blob/main/.github/workflows/trymerge.yml). When the workflow completes successfully the bot closes the PR 

[image](https://github.com/pytorch/pytorch/commit/144fde4fd20ae018028d264441d95857b649f32f)

The changes proposed in this PR was added in `main` in https://github.com/pytorch/pytorch/commit/144fde4fd20ae018028d264441d95857b649f32f.

> And if the code has been effectively merged, on which release would it be available ?

It is in the [release branch for 2.5](https://github.com/pytorch/pytorch/blob/32f585d9346e316e554c8d9bf7548af9f62141fc/aten/src/ATen/autocast_mode.cpp#L72) and I expect it'll be available in 2.5.0.

**Comment by awmartin** [comment]
I get the following warning for autocast on MPS using pytorch 2.6.0 nightly:

```
sam2/.env/lib/python3.11/site-packages/torch/amp/autocast_mode.py:332: UserWarning: In MPS autocast, but the target dtype is not supported. Disabling autocast.
MPS Autocast only supports dtype of torch.bfloat16 currently.
```

My offending code _appears_ to reference the proper dtype:

```python
with torch.inference_mode(), torch.autocast("mps", dtype=torch.bfloat16):
```

However, it seems the actual supported type is `torch.float16` as per [this line](https://github.com/Chao1Han/pytorch/commit/fd35f1ca5af15c7cc46d0b569aac5ef3af58f1f4#diff-0ca6f7e7a58e1b11315cda35388d83cc646f76f14eb1e5ec2d5fd520f80d5a9bR326). Is this expected? Or is the [warning message on line 330](https://github.com/Chao1Han/pytorch/commit/fd35f1ca5af15c7cc46d0b569aac5ef3af58f1f4#diff-0ca6f7e7a58e1b11315cda35388d83cc646f76f14eb1e5ec2d5fd520f80d5a9bR330) incorrect?

**Comment by hvaara** [comment]
I created #139190 to add some context and track a fix. tl;dr: The error message is incorrect and should be updated. Thanks for reporting the issue, @awmartin!

**Comment by hvaara** [comment]
I added a proposal to extend MPS autocast support to bf16 in #139390.

cc @awmartin

**Comment by awmartin** [comment]
@hvaara Much appreciated!

**Pull request closed without merging**